### ISSUE TITLE
Make JS tests run

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#254]
This is a continuation of #247 related to the JS test suite and friends. The primary JS test runner is `mach check-spidermonkey`. Here what it does (in parenthesis — it's `make` equivalent, if any):
```
jit-tests: jit_test.py using js.exe and *.js test scripts (make check-jit-test)
jstests: jstests.py using js.exe and *.js test scripts (make check-jstests)
jsapi-tests: jsapi-tests.exe
check-style: check_spidermonkey_style.py (make check-style)
check-masm: check_macroassembler_style.py (make check-masm)
```

**Comment by dmik** [comment]
On OS/2, most of this runs fine. Current problems are the following:
* jit-tests - one failure (6 variants):
```
byteSizeOfScript(f1) = 144
byteSizeOfScript(f2) = 144
D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js:46:1 Error: Assertion failed: got false, expected true
Stack:
  @D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js:46:1
Exit code: 3
FAIL - heap-analysis/byteSize-of-scripts.js
...
FAILURES:
    D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js
    --ion-eager --ion-offthread-compile=off D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js
    --ion-eager --ion-offthread-compile=off --non-writable-jitcode --ion-check-range-analysis --ion-extra-checks --no-sse3 --no-threads D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js
    --baseline-eager D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js
    --baseline-eager --no-fpu D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js
    --no-baseline --no-ion D:/Coding/mozilla/master/js/src/jit-test/tests/heap-analysis/byteSize-of-scripts.js
TIMEOUTS:
```
Running it in `SET NUMBER_OF_PROCESSORS=N` mode, however (where N > 1) also shows some timeout problems for reasons yet unknown:
```
TIMEOUTS:
    D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f D:/Coding/mozilla/master/js/src/jit-test/lib/prologue.js --js-cache D:/Coding/mozilla/master/js/src/jit-test/.js-cache --no-baseline --no-ion -e "const platform='os2knix'; const libdir='D:/Coding/mozilla/master/js/src/jit-test/lib/'; const scriptdir='D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/'" -f D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/testTimeout-deactivate-reactivate-signals.js
    D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f D:/Coding/mozilla/master/js/src/jit-test/lib/prologue.js --js-cache D:/Coding/mozilla/master/js/src/jit-test/.js-cache --ion-eager --ion-offthread-compile=off -e "const platform='os2knix'; const libdir='D:/Coding/mozilla/master/js/src/jit-test/lib/'; const scriptdir='D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/'" -f D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/testTimeout1-nosignals.js
    D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f D:/Coding/mozilla/master/js/src/jit-test/lib/prologue.js --js-cache D:/Coding/mozilla/master/js/src/jit-test/.js-cache -e "const platform='os2knix'; const libdir='D:/Coding/mozilla/master/js/src/jit-test/lib/'; const scriptdir='D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/'" -f D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/testTimeout1-nosignals.js
    D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f D:/Coding/mozilla/master/js/src/jit-test/lib/prologue.js --js-cache D:/Coding/mozilla/master/js/src/jit-test/.js-cache --ion-extra-checks --non-writable-jitcode --ion-check-range-analysis --ion-offthread-compile=off --no-threads --ion-eager --no-sse3 -e "const platform='os2knix'; const libdir='D:/Coding/mozilla/master/js/src/jit-test/lib/'; const scriptdir='D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/'" -f D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/testTimeout1-nosignals.js
    D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f D:/Coding/mozilla/master/js/src/jit-test/lib/prologue.js --js-cache D:/Coding/mozilla/master/js/src/jit-test/.js-cache --baseline-eager --no-fpu -e "const platform='os2knix'; const libdir='D:/Coding/mozilla/master/js/src/jit-test/lib/'; const scriptdir='D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/'" -f D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/testTimeout1.js
    D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f D:/Coding/mozilla/master/js/src/jit-test/lib/prologue.js --js-cache D:/Coding/mozilla/master/js/src/jit-test/.js-cache --baseline-eager -e "const platform='os2knix'; const libdir='D:/Coding/mozilla/master/js/src/jit-test/lib/'; const scriptdir='D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/'" -f D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/testTimeout1.js
    D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f D:/Coding/mozilla/master/js/src/jit-test/lib/prologue.js --js-cache D:/Coding/mozilla/master/js/src/jit-test/.js-cache --ion-extra-checks --non-writable-jitcode --ion-check-range-analysis --ion-offthread-compile=off --no-threads --ion-eager --no-sse3 -e "const platform='os2knix'; const libdir='D:/Coding/mozilla/master/js/src/jit-test/lib/'; const scriptdir='D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/'" -f D:/Coding/mozilla/master/js/src/jit-test/tests/asm.js/testTimeout1.js
```
* jstests - one failure:
```
## ecma/Array/15.4.3.1-2.js: rc = 0, run time = 0.397
TEST-UNEXPECTED-FAIL | ecma/Array/15.4.3.1-2.js | (args: "")
```
* ~~jsapi-tests - one failure:~~ (fixed within #255)
```
TEST-UNEXPECTED-FAIL | testGCAllocator | CHECK failed: positionIsCorrect("x-ooxxx---------", stagingArea, chunkPool, tempChunks)
D:/Coding/mozilla/master/js/src/jsapi-tests/testGCAllocator.cpp:145:CHECK failed: positionIsCorrect("x-ooxxx---------", stagingArea, chunkPool, tempChunks)
```

**Comment by dmik** [comment]
Regarding the jsapi-tests failure, it seems that kLIBC's `DosAllocMemEx` which allows to allocate a memory block at a given address, doesn't work good enough. Should be fixable. Somehow JS needs this now for proper GC management.

**Comment by dryeo** [comment]
I wonder if one problem is that there tests and JS engine is setup more 
to use jemalloc rather then mozalloc?

**Comment by dmik** [comment]
@dryeo see #255. Neither jemalloc nor mozalloc is used in that part (at least in ESR 49) — OS direct memory primitives are (VirtualAlloc on Win, DosAllocMem on OS/2 and mmap everywhere else). Yes, it's really strange that Mozilla uses so many different dynamic memory allocation helpers but so it is.

**Comment by dmik** [comment]
Updated the `jsapi-tests` status in the first comment. Note that this leaves us with the `jstests` Array timeout and a `byteSize-of-scripts` failure. The latter is definitely not related to #248. The Array timeout's nature is not clear to me so far as it works fine when run separately. Needs a bit of further investigation.

**Comment by dmik** [comment]
A new run of `jstests` doesn't show up any problems with `Array`, instead two failures with the locale pop up again. Something's still wrong with my locale WRT these tests but I postpone it as not relevant to the main problem.
```
## js1_5/extensions/toLocaleFormat-01.js: rc = 0, run time = 0.25
D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f shell.js -f js1_5/shell.js -f js1_5/extensions/shell.js -f js1_5/extensions/toLocaleFormat-01.js
BUGNUMBER: 291494
STATUS: Date.prototype.toLocaleFormat extension
 PASSED! Date.toLocaleFormat("%a")
 PASSED! Date.toLocaleFormat("%A")
 PASSED! Date.toLocaleFormat("%b")
 PASSED! Date.toLocaleFormat("%B")
 PASSED! Date.toLocaleFormat("%d")
 PASSED! Date.toLocaleFormat(%H)
 PASSED! Date.toLocaleFormat(%I)
 PASSED! Date.toLocaleFormat("%j")
 PASSED! Date.toLocaleFormat("%m")
 PASSED! Date.toLocaleFormat("%M")
 FAILED! [reported from test()] Date.toLocaleFormat("%p") is AM or PM : Expected value 'true', Actual value 'false' 
 PASSED! Date.toLocaleFormat("%S")
 PASSED! Date.toLocaleFormat("%U")
 FAILED! [reported from test()] Date.toLocaleFormat("%w") : Expected value '6', Actual value '06' 
 PASSED! Date.toLocaleFormat("%W")
 PASSED! Date.toLocaleFormat("%y")
 PASSED! Date.toLocaleFormat("%Y")
 PASSED! Date.toLocaleFormat("%%")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("%Y %y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
 PASSED! Date.toLocaleFormat("xxxx%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y%Y")
REGRESSION - js1_5/extensions/toLocaleFormat-01.js
D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f shell.js -f js1_5/shell.js -f js1_5/Function/shell.js -f js1_5/Function/10.1.6-01.js
## js1_5/extensions/toLocaleFormat-02.js: rc = 0, run time = 0.214
D:/Coding/mozilla/master-build-test/dist/bin/js.exe -f shell.js -f js1_5/shell.js -f js1_5/extensions/shell.js -f js1_5/extensions/toLocaleFormat-02.js
BUGNUMBER: 291494
STATUS: Date.prototype.toLocaleFormat extension
 FAILED! [reported from test()] Date.toLocaleFormat("%C") : Expected value '20', Actual value '02' 
 FAILED! [reported from test()] Date.toLocaleFormat("%C%y") == Date.toLocaleFormat("%Y") : Expected value '0205', Actual value '2005' 
 PASSED! Date.toLocaleFormat("%m/%d/%y") == Date.toLocaleFormat("%D")
 PASSED! Date.toLocaleFormat("%e")
 PASSED! Date.toLocaleFormat("%Y-%m-%d") == Date.toLocaleFormat("%F")
 FAILED! [reported from test()] Date.toLocaleFormat("%g") : Expected value '05', Actual value 'Sat Jun 04 2005 17:00:00 GMT-0700 (PDT)' 
 FAILED! [reported from test()] Date.toLocaleFormat("%G") : Expected value '2005', Actual value 'Sat Jun 04 2005 17:00:00 GMT-0700 (PDT)' 
 PASSED! Date.toLocaleFormat("%b") == Date.toLocaleFormat("%h")
 PASSED! Date.toLocaleFormat("%n") == "\n"
 PASSED! Date.toLocaleFormat("%I:%M:%S %p") == Date.toLocaleFormat("%r")
 FAILED! [reported from test()] Date.toLocaleFormat("%H:%M") == Date.toLocaleFormat("%R") : Expected value '17:00', Actual value 'Sat Jun 04 2005 17:00:00 GMT-0700 (PDT)' 
 PASSED! Date.toLocaleFormat("%t") == "\t"
 PASSED! Date.toLocaleFormat("%H:%M:%S") == Date.toLocaleFormat("%T")
 FAILED! [reported from test()] Date.toLocaleFormat("%u") : Expected value '6', Actual value 'Sat Jun 04 2005 17:00:00 GMT-0700 (PDT)' 
 FAILED! [reported from test()] Date.toLocaleFormat("%V") : Expected value '22', Actual value 'Sat Jun 04 2005 17:00:00 GMT-0700 (PDT)' 
Note: For Date.toLocaleFormat("%m/%d/%y") == Date.toLocaleFormat("%x") to pass in Windows, the Regional Setting for the short date must be set to mm/dd/yyyy
 PASSED! Date.toLocaleFormat("%m/%d/%Y") == Date.toLocaleFormat("%x")
REGRESSION - js1_5/extensions/toLocaleFormat-02.js
```

**Comment by dmik** [comment]
I'm removing this from ESR 49 GA2 milestone as it doesn't seem that the slowdown can be revealed by the JS test suite. It might not be related to JS at all. As far as the tests cover, JS works quite well on OS/2. We will have to try out other test suites.

**Comment by dmik** [comment]
Note that `mach xpcshell-test` also works now but hangs if using multiple threads (looks the same like `jstests` multithreaded hangs). If I limit it to one thread that many tests run fine but then some complain that they can't run in parallel.

**Comment by dmik** [comment]
BTW, I was too fast saying `xpcshell-test` works. When I limit it to 1 thread, most tests end like this:
```
0:06.63 TEST_START: Thread-1 browser/components/dirprovider/tests/unit/test_bookmark_pref.js
0:06.65 LOG: Thread-1 INFO browser/components/dirprovider/tests/unit/test_bookmark_pref.js failed or timed out, will retry.
```
I need to dig deeper to find out what's going on there.
